crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,328 @@
+//!   fetch request.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Instant,
+};
+
+use bytes::BufMut;
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    transport::Transport,
+    AgentId, BoxFut, K2Result, OpId, SpaceId, Url,
+};
+use tokio::sync::{
+    mpsc::{channel, Receiver, Sender},
+    Mutex,
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    pub parallel_request_count: u8,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 120_000.
+    pub cool_down_interval_ms: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            cool_down_interval_ms: 120_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<dyn Transport>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        let inner = Inner::new(config.cool_down_interval_ms);
+        inner.spawn_fetch_tasks(config.parallel_request_count, transport);
+        Self(inner)
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to set.
+            let ops = &mut self.0.state.lock().await.ops;
+            ops.extend(
+                op_list
+                    .clone()
+                    .into_iter()
+                    .map(|op_id| (op_id.clone(), source.clone())),
+            );
+
+            // Pass ops to fetch tasks.
+            op_list.into_iter().for_each(|op_id| {
+                if let Err(err) =
+                    self.0.fetch_queue_tx.try_send((op_id, source.clone()))
+                {
+                    tracing::warn!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: CoolDownList,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
```

Comment:
There is no reason the receiver side needs to be in your top level class. I'd spawn the tokio threads in your constructor so you don't need this here.

Reply:
Ha, that's how it was before, and it looked a bit cleaner with the separate constructor.

Reply:
Reverted https://github.com/holochain/kitsune2/pull/35/commits/5fd4e84b2ad874141ac96443f738d069fbe5d4f5